### PR TITLE
feat: use @chainsafe/blst directly

### PR DESCRIPTION
**Pull request by wemeetagain**
**Motivation**

In order to do some optimizations, we want to remove the `@chainsafe/bls` abstraction layer. We know we won't be running the beacon node or validator in node _only_. In the future, we may relax this restriction if a usecase presents itself.

**Description**

- Replace all uses of `@chainsafe/bls` throughout the monorepo, with the exception of the light client package
- Notable changes:
  - `.toBytes()` -> `.serialize()`
    - serialize defaults to compressed
    - there is no longer a third param to do validation, it needs to be done afterwards with a `.keyValidate()` or `.sigValidate()`
  - `.fromBytes()` -> `.deserialize()`
  - `Pubkey.aggregate`, `bls.aggregatePubkeys` -> `aggregatePubkeys`
    - same for signatures
  - `Signature.verifyAggregate` -> `fastVerifyAggregate`
  - all verification functions into (msg(s), pubkey(s), sig(s)) order -- <- very important to get right

### CODE REVIEW

File: packages/beacon-node/test/spec/bls/bls.ts [hunk 1]
```diff
@@ -31,7 +40,11 @@ export const testFnByType: Record<string, "skip" | ((data: any) => any)> = {
  */
 function aggregate_verify(input: {pubkeys: string[]; messages: string[]; signature: string}): boolean {
   const {pubkeys, messages, signature} = input;
-  return bls.verifyMultiple(pubkeys.map(fromHexString), messages.map(fromHexString), fromHexString(signature));
+  try {
```
Comment (matthewkeil) [inline packages/beacon-node/test/spec/bls/bls.ts:43]: @wemeetagain I added this try/catch for the spec tests but am updating the base lib to not throw for this so it better meets the spec.

File: packages/beacon-node/test/spec/bls/bls.ts [hunk 2]
```diff
@@ -119,7 +146,11 @@ function sign(input: {privkey: string; message: string}): string | null {
  */
 function verify(input: {pubkey: string; message: string; signature: string}): boolean {
   const {pubkey, message, signature} = input;
-  return bls.verify(fromHexString(pubkey), fromHexString(message), fromHexString(signature));
+  try {
+    return _verify(fromHexString(message), fromHexString(pubkey), fromHexString(signature));
+  } catch {
+    return false;
+  }
```
Comment (matthewkeil) [inline packages/beacon-node/test/spec/bls/bls.ts:151]: @wemeetagain I added this try/catch for the spec tests but am updating the base lib to not throw for this so it better meets the spec.

File: packages/beacon-node/test/spec/general/bls.ts [hunk 3]
```diff
@@ -80,7 +89,11 @@ function aggregate(input: string[]): string {
  */
 function aggregate_verify(input: {pubkeys: string[]; messages: string[]; signature: string}): boolean {
   const {pubkeys, messages, signature} = input;
-  return bls.verifyMultiple(pubkeys.map(fromHexString), messages.map(fromHexString), fromHexString(signature));
+  try {
+    return aggregateVerify(messages.map(fromHexString), pubkeys.map(fromHexString), fromHexString(signature));
+  } catch {
+    return false;
+  }
```
Comment (matthewkeil) [inline packages/beacon-node/test/spec/general/bls.ts:96]: @wemeetagain I added this try/catch for the spec tests but am updating the base lib to not throw for this so it better meets the spec.

File: packages/beacon-node/test/spec/general/bls.ts [hunk 4]
```diff
@@ -169,5 +189,9 @@ function sign(input: {privkey: string; message: string}): string | null {
  */
 function verify(input: {pubkey: string; message: string; signature: string}): boolean {
   const {pubkey, message, signature} = input;
-  return bls.verify(fromHexString(pubkey), fromHexString(message), fromHexString(signature));
+  try {
+    return _verify(fromHexString(message), fromHexString(pubkey), fromHexString(signature));
+  } catch {
+    return false;
+  }
```
Comment (matthewkeil) [inline packages/beacon-node/test/spec/general/bls.ts:195]: @wemeetagain I added this try/catch for the spec tests but am updating the base lib to not throw for this so it better meets the spec.

**Comment by matthewkeil** [comment]
Metrics on feat2-mainnet do not look great.  They should be equal to unstable/stable but seems to be performing worse for some reason... Will try and swap for another feat group tomorrow and see if its the machine

**Comment by matthewkeil** [comment]
After running for a while it seems like this branch is actually fine.  @tuyennhv would love your opinion as well but I'm going to approve

File: packages/beacon-node/src/chain/bls/jobItem.ts [hunk 5]
```diff
@@ -71,11 +76,13 @@ export function jobItemWorkReq(job: JobQueueItem, format: PointFormat, metrics:
       // and not a problem in the near future
       // this is monitored on v1.11.0 https://github.com/ChainSafe/lodestar/pull/5912#issuecomment-1700320307
       const timer = metrics?.blsThreadPool.signatureDeserializationMainThreadDuration.startTimer();
-      const signatures = job.sets.map((set) => bls.Signature.fromBytes(set.signature, CoordType.affine, true));
+      const signatures = job.sets.map((set) => {
+        const sig = Signature.deserialize(set.signature, CoordType.affine);
```
Comment (twoeths) [inline packages/beacon-node/src/chain/bls/jobItem.ts]: it's worth to add `signatureFromBytes()` util vs the current `signatureFromBytesNoCheck()` util

**Review by twoeths — DISMISSED** [review]
Looks good to me, had a suggestion for `signatureFromBytes()` util
also `signatureFromBytesNoCheck()` should be moved to `@lodestar/utils` instead so that both could be consumed by different packages

File: packages/beacon-node/src/chain/bls/jobItem.ts (hunk 5 above)
Comment (matthewkeil) [reply packages/beacon-node/src/chain/bls/jobItem.ts]: Updated.

**Comment by matthewkeil** [comment]
> Looks good to me, had a suggestion for `signatureFromBytes()` util also `signatureFromBytesNoCheck()` should be moved to `@lodestar/utils` instead so that both could be consumed by different packages

Added them both to `@lodestar/utils`

**Comment by matthewkeil** [comment]
@twoeths @philknows Spoke with @wemeetagain and we will merge this one after interop.  Should not be an issue but why add more complexity to an already moving target.  Can sit for the week and get merged when we get back.  All is ready and just waiting for reapproval after the last update for the comment @twoeths had

File: packages/beacon-node/src/chain/bls/singleThread.ts [hunk 6]
```diff
@@ -40,12 +38,14 @@ export class BlsSingleThreadVerifier implements IBlsVerifier {
...
     const timer = this.metrics?.blsThreadPool.mainThreadDurationInThreadPool.startTimer();
-    const pubkey = bls.PublicKey.aggregate(sets.map((set) => set.publicKey));
+    const pubkey = aggregatePublicKeys(sets.map((set) => set.publicKey));
     let isAllValid = true;
     // validate signature = true
     const signatures = sets.map((set) => {
       try {
-        return bls.Signature.fromBytes(set.signature, CoordType.affine, true);
+        const sig = Signature.deserialize(set.signature, CoordType.affine);
+        sig.sigValidate();
```
Comment (twoeths) [inline packages/beacon-node/src/chain/bls/singleThread.ts]: use signatureFromBytes() here

File: packages/beacon-node/test/spec/general/bls.ts [hunk 7]
```diff
@@ -139,9 +152,16 @@ function eth_fast_aggregate_verify(input: {pubkeys: string[]; message: string; s
 function fast_aggregate_verify(input: {pubkeys: string[]; message: string; signature: string}): boolean | null {
   const {pubkeys, message, signature} = input;
   try {
-    return bls.Signature.fromBytes(fromHexString(signature), undefined, true).verifyAggregate(
-      pubkeys.map((hex) => bls.PublicKey.fromBytes(fromHexString(hex), CoordType.jacobian, true)),
-      fromHexString(message)
+    const sig = Signature.deserialize(fromHexString(signature), undefined);
+    sig.sigValidate();
```
Comment (twoeths) [inline packages/beacon-node/test/spec/general/bls.ts]: need to search for all ` sigValidate()` and replace with `signatureFromBytes()`

Comment (matthewkeil) [reply packages/beacon-node/test/spec/general/bls.ts]: Ok. Good idea!

File: packages/beacon-node/src/chain/bls/singleThread.ts (hunk 6 above)
Comment (matthewkeil) [reply packages/beacon-node/src/chain/bls/singleThread.ts]: Updated.

File: packages/beacon-node/test/spec/general/bls.ts (hunk 7 above)
Comment (matthewkeil) [reply packages/beacon-node/test/spec/general/bls.ts]: Updated.

**Comment by wemeetagain** [comment]
:tada: This PR is included in [v1.19.0](https://github.com/ChainSafe/lodestar/releases/tag/v1.19.0) :tada:

**Pull request merged**